downstairs/src/extent.rs

```diff
@@ -295,26 +310,115 @@ impl Extent {
             move_replacement_extent(dir, number as usize, log)?;
         }
 
-        // Pick the format for the downstairs files
+        // We will migrate every read-write extent with a SQLite file present.
+        //
+        // We use the presence of the .db file as a marker to whether the extent
+        // data file has been migrated.
+        //
+        // Remember, the goal is to turn 2-4 files (extent data, .db, .db-wal,
+        // db-shm) into a single file containing
+        //
+        //  - The extent data (same as before, at the beginning of the file)
+        //  - Metadata and encryption context slots (after extent data in the
+        //  - same file)
         //
-        // Right now, this only supports SQLite-flavored Downstairs
-        let inner = {
-            let mut sqlite_path = path.clone();
-            sqlite_path.set_extension("db");
-            if sqlite_path.exists() {
+        // Here's the procedure:
+        //
+        //  - If the .db file is present
+        //      - Truncate it to the length of the extent data. This is a no-op
+        //        normally, but lets us recover from a partial migration (e.g.
+        //        if we wrote the raw context but crashed before deleting the
+        //        .db file)
+        //      - Open the extent using our existing SQLite extent code
+        //      - Using standard extent APIs, find the metadata and encryption
+        //        context for each block. Encode this in the new raw file format
+        //        (as a Vec<u8>)
+        //      - Close the (SQLite) extent
+        //      - Open the extent data file in append mode, and append the new
+        //        raw metadata + block contexts to the end of the file.
+        //      - Close the extent data file and fsync it to disk
+        //      - Delete the .db file
+        //      - fsync the containing directory (to make the deletion durable;
+        //        not sure if this is necessary)
+        //  - At this point, the .db file is not present and the extent data
+        //    file represents a raw extent that has been persisted to disk
+        //      - If the .db-wal file is present, remove it
+        //      - If the .db-shm file is present, remove it
+        //
+        //  It's safe to fail at any point in this procedure; the next time
+        //  `Extent::open` is called, we will restart the migration (if we
+        //  haven't gotten to deleting the `.db` file).
+        let mut has_sqlite = path.with_extension("db").exists();
+        if has_sqlite && !read_only {
+            info!(log, "Migrating extent {number}");
+            // Truncate the file to the length of the extent data
+            {
+                let f =
+                    OpenOptions::new().read(true).write(true).open(&path)?;
+                f.set_len(def.extent_size().value * def.block_size())?;
```

Comment:
I don't think so!  This call to `set_len` should only ever truncate the file down to the raw data size.

The steps are
1) Truncate file
2) Write new metadata and context region
3) `fsync` the file
4) Remove the `.db` file, indicating that the raw file now contains has metadata and context
5) `fsync` the parent directory to make that removal persistent

If we don't make it to step 4 + 5, we'll re-run steps 1-3 (because the `.db` file will still be present).  I believe that steps 1-3 are safe to run repeatedly with crashes at any point: truncating the file is idempotent, and resets the effects of step 2.